src/RowWiseSparseAdagradFused.cc

```diff
@@ -210,18 +231,64 @@ typename ReturnFunctionSignature<indxType, offsetType>::
         vec_reg_t lr_vreg = vec_reg_t(first_available_vec_reg_id);
         ++first_available_vec_reg_id;
 
+        a->vpbroadcastd(epsilon_vreg, epsilon);
+        a->vpbroadcastd(lr_vreg, lr);
+
+        // Reserve vector registers for random buffer generating
+        // S0...S3: global random buffer state
+        // R: generated random number in uint32_t
+        // r0: extracted random byte (uint8_t) shifted to bits[5...13]
+        // r1: temp
+        vec_reg_t R_vreg, S0_vreg, S1_vreg, S2_vreg, S3_vreg, r0_vreg, r1_vreg;
+        if (areWeightsFp16 && use_stochastic_rounding) {
+          R_vreg = vec_reg_t(first_available_vec_reg_id);
+          first_available_vec_reg_id++;
+          S0_vreg = vec_reg_t(first_available_vec_reg_id);
+          first_available_vec_reg_id++;
+          S1_vreg = vec_reg_t(first_available_vec_reg_id);
+          first_available_vec_reg_id++;
+          S2_vreg = vec_reg_t(first_available_vec_reg_id);
+          first_available_vec_reg_id++;
+          S3_vreg = vec_reg_t(first_available_vec_reg_id);
+          first_available_vec_reg_id++;
+          r0_vreg = vec_reg_t(first_available_vec_reg_id);
+          first_available_vec_reg_id++;
+          r1_vreg = vec_reg_t(first_available_vec_reg_id);
+          first_available_vec_reg_id++;
+
+          // Load random buffer for FP16 stochastic rounding
+          if (instSet == inst_set_t::avx2) {
+            a->vmovdqa(S0_vreg.ymm(), x86::dword_ptr(rand_buffer));
+            a->vmovdqa(S1_vreg.ymm(), x86::dword_ptr(rand_buffer,
+                                                 1 * vlen * sizeof(uint32_t)));
+            a->vmovdqa(S2_vreg.ymm(), x86::dword_ptr(rand_buffer,
+                                                 2 * vlen * sizeof(uint32_t)));
+            a->vmovdqa(S3_vreg.ymm(), x86::dword_ptr(rand_buffer,
+                                                 3 * vlen * sizeof(uint32_t)));
+          } else { // AVX512
+            a->vmovdqa32(S0_vreg, x86::dword_ptr(rand_buffer));
```

Comment:
Can we guarantee that rand_buffer is cache line aligned? Also not sure if we need if / else for avx2 here. We can just use vmovups?

Reply:
Yes the rand_buffer is aligned as alignas(64). But I think unlinged load/store is safer. 

AVX2 does not have a store with int32 granularity (there is vpmaskmovd though that needs mask). And vmovups is for float point. But I guess the behavior is pretty much same.

Reply:
Just checked, vmovdqa is faster than vmovups on BDW/HSW/IVB:

**vmovdqa**
Architecture	Latency	Throughput (CPI)
Icelake	7	0.5
Skylake	0	0.5
Broadwell	1	0.25
Haswell	1	0.25
Ivy Bridge	1	0.5

**vmovups**
Architecture	Latency	Throughput (CPI)
Icelake	7	0.5
Skylake	0	0.5
Broadwell	1	0.5
Haswell	1	0.5
Ivy Bridge	1	1

And depending on the micro-architecture, there might be data bypass delays:
"On the Core2, there is an extra latency of one clock cycle when the output of a µop in the
integer unit is used as input for a µop in the floating point unit, or vice versa."
Page 113 of https://www.agner.org/optimize/microarchitecture.pdf

Reply:
This is a great point. Let me check if we have other cases we should use vmovdqa instead of vmovups.

Reply:
I removed such usages in https://github.com/pytorch/FBGEMM/pull/368 for integer code. Still need to land this PR though.